src/cascadia/TerminalConnection/AzureConnection.cpp

```diff
@@ -0,0 +1,478 @@
+    {
+        AzureConnection* const pInstance = (AzureConnection*)lpParameter;
+        return pInstance->_OutputThread();
+    }
+
+    DWORD AzureConnection::_OutputThread()
+    {
+        //Azure authentication starts here:
+        //step 1a: initiate device code flow
+        json::value deviceCodeResponse = _getDeviceCode();
+
+        //step 1b: print the message and store the device code, polling interval and expiry
+        utility::string_t message = deviceCodeResponse.at(L"message").as_string();
+        _outputHandlers(message + L" This code will expire in 15 minutes.\r\n");
+        utility::string_t devCode = deviceCodeResponse.at(L"device_code").as_string();
+        int pollInterval = std::stoi(deviceCodeResponse.at(L"interval").as_string());
+        int expiresIn = std::stoi(deviceCodeResponse.at(L"expires_in").as_string());
+
+        //step 2: wait for user authentication and obtain the access/refresh tokens
+        json::value authenticatedResponse;
+        try
+        {
+            authenticatedResponse = _waitForUser(devCode, pollInterval, expiresIn);
+        }
+        catch (...)
+        {
+            _outputHandlers(winrt::to_hstring("Exit."));
+            return -1;
+        }
+
+        _accessToken = authenticatedResponse.at(L"access_token").as_string();
+        _refreshToken = authenticatedResponse.at(L"refresh_token").as_string();
+
+        //step 3: getting the tenants and the required tenant id
+        json::value tenantsResponse = _getTenants();
+        json::array tenantList = tenantsResponse.at(L"value").as_array();
+
+        if (tenantList.size() == 0)
+        {
+            _outputHandlers(winrt::to_hstring("Could not find any tenants.\r\n"));
+            return -1;
+        }
+        else if (tenantList.size() == 1)
+        {
+            _tenantID = tenantList.at(0).at(L"tenantId").as_string();
+        }
+        else
+        {
+            _maxSize = tenantList.size();
+            for (int i = 0; i < _maxSize; i++)
+            {
+                utility::string_t str = L"Tenant " + std::to_wstring(i) + L": " + tenantList.at(i).at(L"displayName").as_string() + L" (" + tenantList.at(i).at(L"tenantId").as_string() + L")\r\n";
+                _outputHandlers(str);
+            }
+            utility::string_t str = L"Please enter the desired tenant number (a number from 0 to " + std::to_wstring((_maxSize - 1)) + L" inclusive).\r\n";
+            _outputHandlers(str);
+            _needInput = true;
+            // Use a lock to wait for the user to input a valid number
+            std::unique_lock<std::mutex> tenantNumberLock{ _tenantNumMutex };
+            _canProceed.wait(tenantNumberLock, [=]() {
+                return (_tenantNumber >= 0 && _tenantNumber < _maxSize) || _closing.load();
+            });
+            // User might have closed the tab while we waited for input
+            if (_closing.load())
+            {
+                return -1;

```

Comment:
This would be where I'd use a return code like S_FALSE. Which is "I did everything successfully, there just ended up being nothing to do."